Fix vercel deploy

ofersadgat opened the pull request:
Vercel func's cannot be empty. If you're referencing all paths in the directory, you have to use index

natew commented:
@ofersadgat thanks for taking a look, if you want to keep going for a good solution lmk we can put a $100 bounty on it.

ofersadgat commented:
I have updated the code to allow for nested index paths.

Furthermore, it turns out that there wasnt *any* middleware support for nested paths built. The existing code expected there to be a singular middleware at the root and thats it. I have added support for this. There are a few caveats: the runtime being used to run these functions is the vercel edge. This imposes a few requirements on the running code: namely, it cannot require the normal node modules (i.e. url, http, etc.) and it also cannot do dynamic imports. Consequently, I had to rewrite the middleware wrapper to import all the middlewares and then use a key to get the appropriate one. Doing this has also fixed the requirement of a top level middleware.

The main issue I ran into during the implementation of this was not knowing a good way to implement routing in the middleware. The issue here is that there is a singular middleware function that gets run on every request. Therefore, the middleware function needs to look at the request url and determine what the middleware chain needs to be. In order to do this, I used the BuildInfo.manifest.allRoutes to generate a lookup table using the namedRegex and contextKeys. I serialized this table into the middleware wrapper and then test the request against all the namedRegexs. The reason I had to test against all of them was that multiple regexes were matching (namely the not found regex, the folder regex, and the specific route regex). I resolved the conflict by using the one which had the longest middleware chain. If there is a better way of doing routing in the middleware let me know.

ofersadgat commented:
Just for extra tracking, the issue of dynamic imports not working has an open bug in the next repo: https://github.com/vercel/next.js/issues/34685 and the runtime error is:

```
TypeError [ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING]: A dynamic import callback was not specified.
    at importModuleDynamicallyCallback (node:internal/modules/esm/utils:272:9)
    at $ (user-source.js:19:1249)
    at i (user-source.js:19:1624)
    at Object.g (user-source.js:19:1755)
    at user-source.js:39:28
    at At (edge-functions-bridge-runtime.js:6:8337)
    at edge-functions-bridge-runtime.js:7:16822
    at AsyncLocalStorage.run (node:async_hooks:346:14)
    at Jt (edge-functions-bridge-runtime.js:7:2692)
    at edge-functions-bridge-runtime.js:7:16768
```

ofersadgat commented:
Lastly, a better solution to having a singular middleware that runs routing would be to extract the namedRegexes into the config.json. The main issue with that is that I would need to be able to better filter the routes into their components. I.e. if we have /_middleware.tsx and /app/_middleware.tsx, we currently have 3 types of routes: the generic ones: /index.tsx + /app/index.tsx, the file routes: /app/somefile.tsx, and the not found routes. We would have to filter to the generic routes and then we would need to deduplicate the repeated middleware routes. I.e. /app/index should only run /app/_middleware and not /_middleware since the config.json would naturally run the /_middleware first since its first in the chain anyways. The other potential issue I saw is that namedRegexes are technically implemented incorrectly, but work due to js being more leniant with the way regexs work. I.e. '/' needs to be escaped in an actual regex and idk if the vercel matcher will complain that it isnt.
